src/MergeContext.js

```diff
@@ -377,124 +516,224 @@ class MergeContext {
+
+        if (!(await this._checkActive()))
+            return StepResult.Fail();
+
+        if (!this._prMergeable()) {
+            this._logFailedCondition("mergeable");
+            return StepResult.Fail();
         }
 
-        let params = Util.commonParams();
-        params.number = this.prNumber();
-        params.labels = [];
-        params.labels.push(label);
+        if (await this._previousStagingFailed()) {
+            this._logFailedCondition("lack of fresh staging commit with failed checks");
+            return StepResult.Fail();
+        }
 
-        await GH.addLabels(params);
+        const statusChecks = await this._getPrStatuses();
+        if (statusChecks.failed())
+            return StepResult.Fail();
+
+        if (!this._messageValid) {
+            this._logFailedCondition("valid commit message");
+            return StepResult.Fail();
+        }
+
+        if (!this._approval.granted()) {
+            this._logFailedCondition("approved");
+            return StepResult.Fail();
+        }
+
+        if (this._approval.grantedTimeout())
+            return StepResult.Delay(this._approval.delayMs);
+
+        if (!statusChecks.final())
+            return StepResult.Suspend();
+
+        return StepResult.Succeed();
     }
 
-    async _labelFailedDescription(isValid) {
-        if (this._dryRun("labeling on failed description"))
+    // returns filled StepResult object
+    async update() {
+        this._role = "updater";
+        this._messageValid = this._prMessageValid();
+        this._log("messageValid: " + this._messageValid);
+        this._labelFailedDescription();
+
+        this._approval = await this._checkApproval();
+        this._log("checkApproval: " + this._approval);
+        await this._setApprovalStatus(this._prHeadSha());
+        if (this._prState.staged())
+            await this._setApprovalStatus(this._tagSha);
+
+        if (this._approval.grantedTimeout())
+            return StepResult.Delay(this._approval.delayMs);
+
+        return StepResult.Succeed();
+    }
+
+    // Label manipulation methods
+
+    // applies the cached label state to GitHub
+    // and removes staging tag, if needed
+    async _applyLabels() {
+        if (this._dryRun("apply labels"))
             return;
-        const label = Config.failedDescriptionLabel();
-        if (isValid)
-            await this._removeLabelsIf([label]);
-        else
-            await this._addLabel(label);
+
+        this._log(this._labels.toString());
+        if (this._labels)
+            await this._labels.apply();
     }
 
-    async _unlabelPreconditionsChecking() {
-        await this._removeLabelsIf([
-                Config.passedStagingChecksLabel(),
-                Config.waitingStagingChecksLabel()
-                ]);
+    // Atomize PR post-merge cleanup: suspend this PR
+    // until all required operations are complete.
+    async _finalize() {
+        if (!this._prState.postStaged() || this._dryRun("finalize"))
```

Comment:
If this method is not for other states, then we should assert postStaged() instead of silently suspending a PR.